enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,480 @@
+struct bpf_lpm_trie_ip_key {
+    __u32 prefixlen;
+    union {
+        __u8 data[4];
+        __u8 data[16];
+    } u;
+} __packed;
+
+struct rulesStore {
+    __u32 numRules;
+    struct ruleInfo rulesInfo[0];
+} __packed;
+```
+
+- eBPF statistics
+
+Implementation will use perf events map 
+```c
+struct bpf_map_def SEC("maps") rule_stats_map = {
+        .type		= BPF_MAP_TYPE_PERF_EVENT_ARRAY,
+        .key_size	= sizeof(int),
+        .value_size	= sizeof(struct ruleStatistics),
+        .max_entries	= MAX_CPUS,
+};
+
+struct ruleStatistics {
+    __u64 pkts;
+    __u64 bytes;
+} __packed;
+```
+
+the key will be the ruleId and if we need to capture the packet's header for a specific rule we just need
+to modify the flags setting to factor in how many bytes from the packet's header we need to capture for example :-
+
+```
+#define MAX_HEADER_SIZE 512
+__u64 flags = BPF_F_CURRENT_CPU;
+
+flags |= (__u64)MAX_HEADER_SIZE << 32;
+```
+### User Stories
+
+- Design ingress node Firewall CRD and APIs.
+
+- Design and Implement ingress node firewall CNO controller to manage new CRD "reconciler" . 
+  
+- Implement eBPF tabels loader and attach to the selected interfaces.
+
+- Implement implicit firewall pinholes rules to allow fail safe in case of invalid configuration (user shoudln't try to overwrite those rules).
+
+- Implement verification webhook to validate configuration and protect against invalid rules.
+  
+- Create Ingress node eBPF OCP repro.
+
+- Implement and build Ingress node eBPF kernel hook to do ingress node firewall.
+
+- Ingress node firewall E2E testing.
+
+### API Extensions
+
+A new cluster scoped CRD is introduced to support this feature this CRD will be managed by `ClusterNetworkOperator`
+```golang
+type IngressNodeFirewallFromCIDR struct {
+	// FromCIDR is the IPv4/IPv6 CIDR from which we apply node firewall rule
+	// +kubebuilder:validation:one-of[]=Type=string;Format:ipv4;Format:ipv6
+	FromCIDR string `json:"fromCIDR" protobuf:"bytes,1,opt,name=fromCIDR"`
+}
+
+// IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols
+type IngressNodeFirewallICMPRule struct {
+	// ICMPType define ICMP Type Numbers (RFC 792).
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=43
+	// +kubebuilder:validation:Minimum:=0
+	ICMPType int8 `json:"icmpType" protobuf:"varint,1,opt,name=icmpType"`
+
+	// ICMPCode define ICMP Code ID (RFC 792).
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=16
+	// +kubebuilder:validation:Minimum:=0
+	ICMPCode int8 `json:"icmpCode" protobuf:"varint,2,opt,name=icmpCode"`
+}
+
+// IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols
+type IngressNodeFirewallProtoRule struct {
+    // start-end range of dstPorts or [port1, port2,..., portn].
+    // +kubebuilder:validation:Optional
+    // +optional
+    Ports []int32 `json:"ports" protobuf:"varint,2,opt,name=ports"`
```

Comment:
You should have min and max values.  To do this with a slice, you might need to define a new type and specify min and max on the type:
```go
// Port is a TCP, UDP, or SCTP port number.
//
// +kubebuilder:validation:Minimum:=1
// +kubebuilder:validation:Maximum:=65535
type Port int32

type IngressNodeFirewallProtoRule struct {
    // [...]
    Ports []Port `json:"ports" protobuf:"varint,2,opt,name=ports"`
```

Reply:
I am going with single port per protocol ran into issues with XDP code having nested loops will add it to none-goal and see how people will react to it